cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs

```diff
@@ -0,0 +1,741 @@
+          Right script -> Right script
+
+  where
+    -- TODO: Think of a way to get type checker to warn when there is a missing
+    -- script version.
+    textEnvTypes :: [FromSomeType HasTextEnvelope ScriptInAnyLang]
+    textEnvTypes =
+      [ FromSomeType (AsScript AsSimpleScriptV1)
+                     (ScriptInAnyLang (SimpleScriptLanguage SimpleScriptV1))
+
+      , FromSomeType (AsScript AsSimpleScriptV2)
+                     (ScriptInAnyLang (SimpleScriptLanguage SimpleScriptV2))
+
+      , FromSomeType (AsScript AsPlutusScriptV1)
+                     (ScriptInAnyLang (PlutusScriptLanguage PlutusScriptV1))
+
+      , FromSomeType (AsScript AsPlutusScriptV2)
+                     (ScriptInAnyLang (PlutusScriptLanguage PlutusScriptV2))
+      ]
+
+    toMinimumSimpleScriptVersion :: SimpleScript SimpleScriptV2
+                                 -> ScriptInAnyLang
+    toMinimumSimpleScriptVersion s =
+      -- TODO alonzo: this will need to be adjusted when more versions are added
+      -- with appropriate helper functions it can probably be done in an
+      -- era-generic style
+      case adjustSimpleScriptVersion SimpleScriptV1 s of
+        Nothing -> ScriptInAnyLang (SimpleScriptLanguage SimpleScriptV2)
+                                   (SimpleScript SimpleScriptV2 s)
+        Just s' -> ScriptInAnyLang (SimpleScriptLanguage SimpleScriptV1)
+                                   (SimpleScript SimpleScriptV1 s')
+
+
+-- Tx & TxBody
+
+newtype CddlTx = CddlTx {unCddlTx :: InAnyCardanoEra Tx} deriving (Show, Eq)
+
+readFileTx :: FilePath -> IO (Either CddlError (InAnyCardanoEra Tx))
+readFileTx fp = do
+  eAnyTx <- readFileInAnyCardanoEra AsTx fp
+  case eAnyTx of
+    Left e -> fmap unCddlTx <$> acceptTxCDDLSerialisation e
+    Right tx -> return $ Right tx
+
+-- IncompleteCddlFormattedTx is an CDDL formatted tx or partial tx
+-- (respectively needs additional witnesses or totally unwitnessed)
+-- while UnwitnessedCliFormattedTxBody is CLI formatted TxBody and
+-- needs to be key witnessed.
+
+data IncompleteTx
+  = UnwitnessedCliFormattedTxBody (InAnyCardanoEra TxBody)
+  | IncompleteCddlFormattedTx (InAnyCardanoEra Tx)
+
+readFileTxBody :: FilePath -> IO (Either CddlError IncompleteTx)
+readFileTxBody fp = do
+  eTxBody <- readFileInAnyCardanoEra AsTxBody fp
+  case eTxBody of
+    Left e -> fmap (IncompleteCddlFormattedTx . unCddlTx) <$> acceptTxCDDLSerialisation e
+    Right txBody -> return $ Right $ UnwitnessedCliFormattedTxBody txBody
+
+data CddlError = CddlErrorTextEnv
+                   !(FileError TextEnvelopeError)
+                   !(FileError TextEnvelopeCddlError)
+               | CddlIOError (FileError TextEnvelopeError)
+
+renderCddlError :: CddlError -> Text
+renderCddlError (CddlErrorTextEnv textEnvErr cddlErr) =
+  "Failed to decode neither the cli's serialisation format nor the ledger's \
+  \CDDL serialisation format. TextEnvelope error: " <> Text.pack (displayError textEnvErr) <> "\n" <>
```

Comment:
Prefer to not use multiline strings.  Prefer using `<>`.

Reply:
Yeah fair enough. We should purge the use of `\`. Can you open an issue for this?

Reply:
https://github.com/input-output-hk/cardano-node/issues/4585